### PR TITLE
feat: implement coincap market service (fixes #281)

### LINKED ISSUE #281: CoinCap MarketService Implementation
## Background
Our new web application (app.shapeshift.com) relies on a MarketService for its market data.  Currently, we have a single implementation of this MarketService abstraction using coingecko. 

We would like to create a second implementation that uses https://coincap.io as the backend.

The interface to implement can be found in here the lib repo:
https://github.com/shapeshift/lib/blob/b9d28ebda8896d50ce99e7804f277fe8bd491391/packages/market-service/src/api.ts#L3

And this is the Coingecko implementation being instantiated as the sole implementation:
https://github.com/shapeshift/lib/blob/b9d28ebda8896d50ce99e7804f277fe8bd491391/packages/market-service/src/index.ts#L15

Finally this is the Coingecko implementation itself which can be used as a reference:
https://github.com/shapeshift/lib/tree/main/packages/market-service/src/coingecko

## Acceptance Criteria
- Create a new implementation of the MarketService interface for Coincap 
- Implement full test suite for the added functionality modeled after [coingecko.test.ts](https://github.com/shapeshift/lib/blob/b9d28ebda8896d50ce99e7804f277fe8bd491391/packages/market-service/src/coingecko/coingecko.test.ts)

## References
- https://docs.coincap.io/

### PR DESCRIPTION
**Pull request by Zerquix18**
Hello. This is the implementation of the CoinCap MarketService.

It follows the interface for MarketService and contains unit tests, all should be successful.

Two things to notice:
1. In the unit test, I'm directly instancing `CoinCapService` so that the mocks work. Since Coingecko is first in the list of providers, calling the functions in the parent directory causes a problem with the mocks.
2. I think coincap doesn't support smart contracts, so the coincapservice will throw an error if the CAIP19 is for a smart contract.

Tried to follow your conventions and make the code a similar to the coingecko service as possible
Let me know what you think :)

### CODE REVIEW

File: packages/market-service/src/coincap/coincap.ts [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    } catch (e) {
+      return {}
+    }
+  }
+
+  findByCaip19 = async ({ caip19 }: MarketDataArgs): Promise<MarketData> => {
+    try {
+      const { tokenId } = fromCAIP19(caip19)
+      const isToken = !!tokenId
+      if (isToken) {
```
Comment (0xdef1cafe) [inline packages/market-service/src/coincap/coincap.ts:72]: this confused me at first but then i looked at the coincap api. 

we don't need to pull the tokenId from the caip19 at all, so we can remove this check and throw here.

see my above comment about implementing a coincap caip adapter, so we can turn unique asset ids (caip19s) into coincap ids (strings), then we can just fetch them by coincap id.

File: packages/market-service/src/coincap/coincap.ts [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+      throw new Error('MarketService(findByCaip19): error fetching market data')
+    }
+  }
+
+  findPriceHistoryByCaip19 = async ({
+    caip19,
+    timeframe
+  }: PriceHistoryArgs): Promise<HistoryData[]> => {
+    const { tokenId } = fromCAIP19(caip19)
+    const id = tokenId ? 'ethereum' : adapters.CAIP19ToCoingecko(caip19)
```
Comment (0xdef1cafe) [inline packages/market-service/src/coincap/coincap.ts:100]: change to coincap adapter

File: packages/market-service/src/coincap/generateMarketCapData.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import fs from 'fs'
+
+import { findAll } from '..'
+
+const generateMarketCapData = async () => {
```
Comment (0xdef1cafe) [inline packages/market-service/src/coincap/generateMarketCapData.ts:5]: we shouldn't store and cache market cap data like we do for assets.

asset data is essentially static so we can generate it offline periodically. market data changes constantly so we can't cache it statically.

File: packages/market-service/src/coincap/coincap.test.ts [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+      mockedAxios.get.mockResolvedValue({ data: [btc] })
+      const pages = 1
+      const perPage = 10
+      await coinMarketService.findAll({ pages, perPage })
+      expect(mockedAxios.get).toHaveBeenCalledTimes(1)
+      const url = 'https://api.coincap.io/v2/assets?limit=10&offset=1'
+      expect(mockedAxios.get).toBeCalledWith(url)
+    })
+
+    it('can map coingecko to caip ids', async () => {
```
Comment (0xdef1cafe) [inline packages/market-service/src/coincap/coincap.test.ts:92]: this needs to be for coincap

**Review by 0xdef1cafe — CHANGES_REQUESTED** [review]

**Comment by 0xdef1cafe** [comment]
hey @Zerquix18 thank you for the PR!

this is a great start, but i added some additional context of what needs to be done to get this working. 

happy with what you did with the unit tests, that's the correct way to go about it.

the main thing is implementing the coincap caip adapter so we know how to talk to coincap correctly. unless @natven can confirm the ids are identical this won't work. in the case that they are, we should at least document it.

also, you'll need to lowercase your PR title, then push up a new commit. we use the angular preset for conventional commits https://github.com/conventional-changelog/conventional-changelog/tree/master/packages/conventional-changelog-angular

i think you can do `echo "feat: my commit message" | npx commitlint` locally in the root of `lib` to check without pushing.

**Comment by Zerquix18** [comment]
Hi @0xdef1cafe 
Thanks for your feedback! I'll be going through it

The only thing I didn't quite understand was the commit title. Do I have to change the commit history and submit a new PR?

**Comment by 0xdef1cafe** [comment]
> Hi @0xdef1cafe Thanks for your feedback! I'll be going through it
> 
> The only thing I didn't quite understand was the commit title. Do I have to change the commit history and submit a new PR?

hey - no need for a new PR. if you just change the title to

`feat: implement coincap market service`

and push any commit (which you will with your changes) it'll work fine :)

looking forward to seeing the changes

File: packages/market-service/src/coincap/generateMarketCapData.ts (hunk 3 above)
Comment (Zerquix18) [reply packages/market-service/src/coincap/generateMarketCapData.ts:5]: I did this following the coingecko example, can remove if needed

**Comment by Zerquix18** [comment]
@0xdef1cafe changed the title and added the adapter for coincap. All tests are passing.
Take a look when you have a chance :smile:

File: packages/market-service/src/coincap/cachedMarketCapData.json [hunk 5]
```diff
@@ -0,0 +1 @@
....06734},"eip155:1/erc20:0xb4fbed161bebcb37afb1cb4a6f7ca18b977ccb25":{"price":"42.97","marketCap":"655726","volume":"2830.58","changePercent24Hr":-1.12718},"eip155:1/erc20:0x68aa3f232da9bdc2343465545794ef3eea5209bd":{"price":"0.00365401","marketCap":"654409","volume":"0","changePercent24Hr":null},"eip155:1/erc20:0xc9f1016d336ef77aee75fc11ad64c5ecf9121332":{"price":"0.00535542","marketCap":"651418","volume":"2193.81","changePercent24Hr":-0.26408},"eip155:1/erc20:0x469eda64aed3a3ad6f868c44564291aa415cb1d9":{"price":"0.655277","marketCap":"651030","volume":"4112.3","changePercent24Hr":-13.28456},"eip155:1/erc20:0x36905fc93280f52362a1cbab151f25dc46742fb5":{"price":"0.00119074","marketCap":"650131","volume":"54276","changePercent24Hr":-6.71093},"eip155:1/erc20:0x461733c17b0755ca5649b6db08b3e213fcf22546":{"price":"0.0075055","marketCap":"649494","volume":"0.8181","changePercent24Hr":-0.21702},"eip155:1/erc20:0x5f474906637bdcda05f29c74653f6962bb0f8eda":{"price":"0.02774396","marketCap":"648755","volume":"6921.09","changePercent24Hr":2.20414},"eip155:1/erc20:0x740623d2c797b7d8d1ecb98e9b4afcf99ec31e14":{"price":"0.451073","marketCap":"646494","volume":"18226.42","changePercent24Hr":1.84352}}
```
Comment (0xdef1cafe) [inline packages/market-service/src/coincap/cachedMarketCapData.json]: delete this

File: packages/market-service/src/coincap/generateMarketCapData.ts (hunk 3 above)
Comment (0xdef1cafe) [reply packages/market-service/src/coincap/generateMarketCapData.ts:5]: ya, please go ahead and delete this file, the generated data, the generate script, and the scripts in package.json - i don't know why i wrote this initially :\

**Review by 0xdef1cafe — COMMENTED** [review]
please see comments

**Comment by 0xdef1cafe** [comment]
@Zerquix18  - hey, this is looking really good, few small cleanup changes requested and we can get this merged pretty soon

**Comment by 0xdef1cafe** [comment]
@Zerquix18 also lint is failing CI - run `yarn lint:fix` from the root to fix things automatically.

**Comment by Zerquix18** [comment]
@0xdef1cafe Removed the scripts and ran the lint:fix command

**Review by 0xdef1cafe — APPROVED** [review]

**Comment by 0xdef1cafe** [comment]
@Zerquix18 looks great, just gonna get one more approval from someone else in the team and we'll get this merged

**Review by 0xean — APPROVED** [review]

**Review by theoboldfrazier — APPROVED** [review]

**Pull request merged**